pkg/lib/cockpit.d.ts

```diff
@@ -192,11 +192,26 @@ declare module 'cockpit' {
 
     /* === String helpers ======================== */
 
+    type FormatOptions = {
+        precision?: number;
+        separate?: boolean;
+    };
+
+    type ByteUnit = {
+        name: string | null;
+        factor: number;
+        selected?: boolean;
+    };
+
     function gettext(message: string): string;
     function gettext(context: string, message?: string): string;
     function ngettext(message1: string, messageN: string, n: number): string;
     function ngettext(context: string, message1: string, messageN: string, n: number): string;
 
     function format_bytes(n: number): string;
     function format(format_string: string, ...args: unknown[]): string;
+    function format_number(n: number, precision?: number): string
+    function format_bytes_per_sec(n: number, factor: number, options?: FormatOptions): string;
+    function format_bits_per_sec(n: number, factor: number, options?: FormatOptions): string;
+    function get_byte_units(guide_value: number, factor: value): ByteUnit[];
```

Comment:
Looks like `factor` is optional.

Reply:
...and, if specified, the only meaningful values are literally `1000` and `1024`.

Reply:
ditto all of the functions that take `factor`...

Reply:
ack, let's do that!

Reply:
Done and fixed the return, but this does not end the misery:
```
              const [total_fmt, unit] = cockpit.format_bytes(this.state.memTotal, 1024, { separate: true, precision: 2 });
E             const used_fmt = cockpit.format_bytes(this.samples[3], unit, { separate: true, precision: 2 })[0];     ■ No overload matches this call.    Overload 1 of 2, '(n: number, factor?: 1000 | 1024 | undefined, options?: (FormatOptions & { separate?: false | undefined; }) | undefined): string', gave the following error.      Argument of
```

So all these format functions accept the factor as string which is annoying! Even weirder it seems to convert `Mib` back to 1024 or MB to 1000. That is magic! PURE MAGIC!

@martinpitt how would you feel about breaking this feature? I think we should type the "correct" API.

Reply:
Note there is also backwards compatibility for this:
```
    // backwards compatible API for separate flag
    assert.deepEqual(cockpit.format_bytes_per_sec(2555, 1024, true),
                     ["2.50", "KiB/s"]);
```
I also rather not type this.

Reply:
@jelly I'm all for cleaning up the API, but not at the expense of breaking existing code. TBH I've never seen calls that pass it as a string or as "Mib", but remember that this is a runtime dependency and a very popular function -- please grep existing cockpit projects (also external ones like cockpit-navigator) for existing "bad/smelly" ways of calling that. I guess/hope there are none, but we need to be sure.

Reply:
See:
```
              const [total_fmt, unit] = cockpit.format_bytes(this.state.memTotal, 1024, { separate: true, precision: 2 });
             const used_fmt = cockpit.format_bytes(this.samples[3], unit, { separate: true, precision: 2 })[0];  
```
`unit` should be a string, so now typescript goes no no no no no :)

I am indeed not happy about breaking it, we should investigate all these code paths. But yeah this smells :)

So that's why I say we type "incorrectly" with the preference to type the correct API.